### ISSUE TITLE
Error while compiling async_ssl.v0.17.0

### ISSUE THREAD

**Issue by yrk-pub** [janestreet/async_ssl#39]
During installation on a 5.1.0 switch using dune I get this error message: 

```
# [...]
# /usr/include/openssl/bn.h:346:24: note: expected ‘BIGNUM **’ {aka ‘struct bignum_st **’} but argument is of type ‘void **’
#   346 | int BN_hex2bn(BIGNUM **a, const char *str);
#       |               ~~~~~~~~~^
# ffi_generated_stubs.c: In function ‘async_ssl_stub_86_SSL_get0_alpn_selected’:
# ffi_generated_stubs.c:574:33: error: passing argument 2 of ‘SSL_get0_alpn_selected’ from incompatible pointer type [-Wincompatible-pointer-types]
#   574 |    SSL_get0_alpn_selected(x397, x398, x399);
#       |                                 ^~~~
#       |                                 |
#       |                                 char **
# /usr/include/openssl/ssl.h:850:67: note: expected ‘const unsigned char **’ but argument is of type ‘char **’
#   850 | void SSL_get0_alpn_selected(const SSL *ssl, const unsigned char **data,
#       |
```                           
                       

OpenSSL on my system is at 3.3.0

**Comment by vbmithr** [comment]
Same here.

**Comment by DarioHett** [comment]
Same here, tried with 5.1.1 and 5.2.0, as well as versions 0.16.0, 0.17.0 and 0.15.0.

dnf info libffi libffi-devel openssl --installed
Installed Packages
Name         : libffi
Version      : 3.4.4
Release      : 7.fc40
Architecture : i686
Size         : 77 k
Source       : libffi-3.4.4-7.fc40.src.rpm
Repository   : @System
From repo    : fedora
Summary      : A portable foreign function interface library
URL          : http://sourceware.org/libffi
License      : MIT
Description  : [...] 

Name         : libffi
Version      : 3.4.4
Release      : 7.fc40
Architecture : x86_64
Size         : 82 k
Source       : libffi-3.4.4-7.fc40.src.rpm
Repository   : @System
From repo    : fedora
Summary      : A portable foreign function interface library
URL          : http://sourceware.org/libffi
License      : MIT
Description  : [...]

Name         : libffi-devel
Version      : 3.4.4
Release      : 7.fc40
Architecture : x86_64
Size         : 33 k
Source       : libffi-3.4.4-7.fc40.src.rpm
Repository   : @System
From repo    : fedora
Summary      : Development files for libffi
URL          : http://sourceware.org/libffi
License      : MIT
Description  : The libffi-devel package contains libraries and header files for
             : developing applications that use libffi.

Name         : openssl
Epoch        : 1
Version      : 3.2.1
Release      : 2.fc40
Architecture : x86_64
Size         : 1.7 M
Source       : openssl-3.2.1-2.fc40.src.rpm
Repository   : @System
From repo    : fedora
Summary      : Utilities from the general purpose cryptography library with TLS implementation
URL          : http://www.openssl.org/
License      : Apache-2.0
Description  : The OpenSSL toolkit provides support for secure communications between
             : machines. OpenSSL includes a certificate management tool and shared
             : libraries which provide various cryptographic algorithms and
             : protocols.

**Comment by vbmithr** [comment]
I have a temporary fix here: https://github.com/vbmithr/async_ssl/commit/5206b3b8490cd6f535ae1f6f290f418d2dc70c71

**Comment by DarioHett** [comment]
```sh
opam pin async_ssl https://github.com/vbmithr/async_ssl.git\#5206b3b
[...]
Do you want to continue? [Y/n] Y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
⬇ retrieved async_ssl.v0.17.0  (no changes)
∗ installed async_ssl.v0.17.0
Done.
```

Thank you so much!

**Comment by SquidDev** [comment]
The fix in 77910a421d99c7ccca173b9ec1b6b1b728ab2f82 does not appear to be correct. `-Wincompatible-pointer-types` is not in dune's standard flag list, so removing it does not work! Instead, `-Wno-incompatible-pointer-types` needs to be passed directly:

```diff
diff --git a/src/dune b/src/dune
index bf67dbc..edfe18f 100644
--- a/src/dune
+++ b/src/dune
@@ -24,11 +24,12 @@
   (language c)
   (names ffi_generated_stubs)
   (flags
-   (:standard \ -Werror -pedantic -Wall -Wunused -Wincompatible-pointer-types)
+   (:standard \ -Werror -pedantic -Wall -Wunused)
    -w
    (:include ../bindings/openssl-ccopt.sexp)
    -Wno-int-conversion
-   -Wno-implicit-function-declaration))
+   -Wno-implicit-function-declaration
+   -Wno-incompatible-pointer-types))
  (name async_ssl)
  (public_name async_ssl)
  (modules import version opt verify_mode ffi_generated_types ffi_generated
```
